File: Runtime/STSimTransformer.Ages.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.StochasticTime

Partial Class STSimTransformer

    Private Sub ChangeCellAgeForProbabilisticTransition(
        ByVal simulationCell As Cell,
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal transition As Transition)

        simulationCell.Age = Me.DetermineTargetAgeProbabilistic(
            simulationCell.Age,
            simulationCell.StratumId,
            simulationCell.StateClassId,
            iteration,
            timestep,
            transition)

    End Sub

    Public Function DetermineTargetAgeProbabilistic(
        ByVal currentCellAge As Integer,
        ByVal destinationStratumId As Nullable(Of Integer),
        ByVal destinationStateClassId As Integer,
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal transition As Transition) As Integer

        Return Me.InternalDetermineTargetAgeProbabilistic(
            currentCellAge,
            destinationStratumId,
            destinationStateClassId,
            iteration,
            timestep,
            transition)

    End Function

    Private Function InternalDetermineTargetAgeProbabilistic(
        ByVal currentCellAge As Integer,
        ByVal destinationStratumId As Nullable(Of Integer),
        ByVal destinationStateClassId As Integer,
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal transition As Transition) As Integer

        Dim dt As DeterministicTransition = Nothing

        Me.GetDeterministicTransition(destinationStratumId, destinationStateClassId, iteration, timestep)

        Dim AgeMin As Integer
        Dim AgeMax As Integer

        If (dt Is Nothing) Then
            AgeMin = 0
            AgeMax = Integer.MaxValue
        Else
            AgeMin = dt.AgeMinimum
            AgeMax = dt.AgeMaximum
        End If

        Dim NewAge As Integer

        If Not (transition.AgeReset) Then

            Dim TargetSimCellAge As Integer = (currentCellAge + transition.AgeRelative)

            If (TargetSimCellAge < AgeMin) Then
                NewAge = AgeMin
            ElseIf (TargetSimCellAge > AgeMax) Then
                NewAge = AgeMax
            Else
                NewAge = TargetSimCellAge
            End If

        Else

            Dim TargetSimCellAge As Integer = Math.Max(AgeMin, AgeMin + transition.AgeRelative)

            If (TargetSimCellAge > AgeMax) Then
                TargetSimCellAge = AgeMax
            End If

            NewAge = TargetSimCellAge

        End If

        If (NewAge < 0) Then
            NewAge = 0
        End If

        Return NewAge

    End Function

    Private Sub InitializeCellAge(
        ByVal simulationCell As Cell,
        ByVal stratumId As Integer,
        ByVal stateClassId As Integer,
        ByVal minimumAge As Integer,
        ByVal maximumAge As Integer,
        ByVal iteration As Integer,
        ByVal timestep As Integer)

        Dim dt As DeterministicTransition = Nothing

        If (Me.m_DeterministicTransitionMap IsNot Nothing) Then
            dt = Me.GetDeterministicTransition(stratumId, stateClassId, iteration, timestep)
        End If

        Debug.Assert(minimumAge <> Integer.MaxValue)

        If (dt Is Nothing) Then

            If maximumAge = Integer.MaxValue Then
                simulationCell.Age = minimumAge
            Else
                simulationCell.Age = Me.m_RandomGenerator.GetNextInteger(minimumAge, maximumAge + 1)
            End If

        Else

            Dim AgeMinOut As Integer = 0
            Dim AgeMaxOut As Integer = 0

            GetAgeMinMax(dt, minimumAge, maximumAge, AgeMinOut, AgeMaxOut)

            If AgeMaxOut = Integer.MaxValue Then
                simulationCell.Age = AgeMinOut
            Else
                simulationCell.Age = Me.m_RandomGenerator.GetNextInteger(AgeMinOut, AgeMaxOut + 1)
            End If

        End If

    End Sub

    ''' <summary>
    ''' Initializes the age reporting helper
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub InitializeAgeReportingHelper()

        Debug.Assert(Me.m_AgeReportingHelper Is Nothing)

        Me.m_AgeReportingHelper = New AgeHelper(False, 0, 0)
        Dim dr As DataRow = Me.Project.GetDataSheet(DATASHEET_AGE_TYPE_NAME).GetDataRow()

        If (dr Is Nothing) Then
            Return
        End If

        If (dr(DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME) IsNot DBNull.Value) Then

            If (dr(DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME) Is DBNull.Value) Then

                Me.AddStatusRecord(StatusRecordType.Warning,
                    "Age reporting freqency set without age reporting maximum.  Not reporting ages.")

                Return

            End If

        End If

        If (dr(DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME) IsNot DBNull.Value) Then

            If (dr(DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME) Is DBNull.Value) Then

                Me.AddStatusRecord(StatusRecordType.Warning,
                    "Age reporting maximum set without age reporting frequency.  Not reporting ages.")

                Return

            End If

        End If

        Dim f As Integer = CInt(dr(DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME))
        Dim m As Integer = CInt(dr(DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME))

        If (m < f) Then

            Me.AddStatusRecord(StatusRecordType.Warning,
                "Age reporting maximum is less than age reporting frequency.  Not reporting ages.")

            Return

        End If

        Me.m_AgeReportingHelper = New AgeHelper(True, f, m)

    End Sub

    ''' <summary>
    ''' Gets the age min and age max for the specified deterministic transition and age values
    ''' </summary>
    ''' <param name="dt">The deterministic transition</param>
    ''' <param name="ageMinIn">The minimum age IN value</param>
    ''' <param name="ageMaxIn">The maximum age IN value</param>
    ''' <param name="ageMinOut">The minimum age OUT value</param>
    ''' <param name="ageMaxOut">The maximum age OUT value</param>
    ''' <remarks></remarks>
    Friend Shared Sub GetAgeMinMax(
        ByVal dt As DeterministicTransition,
        ByVal ageMinIn As Integer,
        ByVal ageMaxIn As Integer,
        ByRef ageMinOut As Integer,
        ByRef ageMaxOut As Integer)

        'Normalize
        Dim dtagemin As Integer = Math.Min(dt.AgeMinimum, dt.AgeMaximum)
        Dim dtagemax As Integer = Math.Max(dt.AgeMinimum, dt.AgeMaximum)

        'This should already be normalized
        Debug.Assert(ageMinIn <= ageMaxIn)

        'If any age value is outside of the dt value, set the age value equal to the nearest dt value.
        'For example, if ageMinIn > dt.agemax the set ageMinIn = dt.agemax.

        If (ageMinIn < dtagemin) Then
            ageMinIn = dtagemin
        End If

        If (ageMaxIn < dtagemin) Then
            ageMaxIn = dtagemin
        End If

        If (ageMinIn > dtagemax) Then
            ageMinIn = dtagemax
        End If

        If (ageMaxIn > dtagemax) Then
            ageMaxIn = dtagemax
        End If

        ageMinOut = ageMinIn
        ageMaxOut = ageMaxIn

        Debug.Assert(ageMinOut <= ageMaxOut)

    End Sub

End Class

File: Runtime/STSimTransformer.Ages.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using SyncroSim.Core;
using System.Diagnostics;
using System.Globalization;

namespace SyncroSim.STSim
{
    public partial class STSimTransformer
    {
        private void ChangeCellAgeForProbabilisticTransition(Cell simulationCell, int iteration, int timestep, Transition transition)
        {
            simulationCell.Age = this.DetermineTargetAgeProbabilistic(
                simulationCell.Age, 
                simulationCell.StratumId, 
                simulationCell.StateClassId, 
                iteration, 
                timestep, 
                transition);
        }

        public int DetermineTargetAgeProbabilistic(int currentCellAge, int? destinationStratumId, int destinationStateClassId, int iteration, int timestep, Transition transition)
        {
            return this.InternalDetermineTargetAgeProbabilistic(
                currentCellAge, 
                destinationStratumId, 
                destinationStateClassId, 
                iteration, 
                timestep, 
                transition);
        }

        private int InternalDetermineTargetAgeProbabilistic(int currentCellAge, int? destinationStratumId, int destinationStateClassId, int iteration, int timestep, Transition transition)
        {
            DeterministicTransition dt = this.GetDeterministicTransition(destinationStratumId, destinationStateClassId, iteration, timestep);

            int AgeMin = 0;
            int AgeMax = 0;

            if (dt == null)
            {
                AgeMin = 0;
                AgeMax = int.MaxValue;
            }
            else
            {
                AgeMin = dt.AgeMinimum;
                AgeMax = dt.AgeMaximum;
            }

            int NewAge = 0;

            if (!(transition.AgeReset))
            {
                int TargetSimCellAge = (currentCellAge + transition.AgeRelative);

                if (TargetSimCellAge < AgeMin)
                {
                    NewAge = AgeMin;
                }
                else if (TargetSimCellAge > AgeMax)
                {
                    NewAge = AgeMax;
                }
                else
                {
                    NewAge = TargetSimCellAge;
                }
            }
            else
            {
                int TargetSimCellAge = Math.Max(AgeMin, AgeMin + transition.AgeRelative);

                if (TargetSimCellAge > AgeMax)
                {
                    TargetSimCellAge = AgeMax;
                }

                NewAge = TargetSimCellAge;
            }

            if (NewAge < 0)
            {
                NewAge = 0;
            }

            return NewAge;
        }

        private void InitializeCellAge(Cell simulationCell, int stratumId, int stateClassId, int minimumAge, int maximumAge, int iteration, int timestep)
        {
            DeterministicTransition dt = null;

            if (this.m_DeterministicTransitionMap != null)
            {
                dt = this.GetDeterministicTransition(stratumId, stateClassId, iteration, timestep);
            }

            Debug.Assert(minimumAge != int.MaxValue);

            if (dt == null)
            {
                if (maximumAge == int.MaxValue)
                {
                    simulationCell.Age = minimumAge;
                }
                else
                {
                    simulationCell.Age = this.m_RandomGenerator.GetNextInteger(minimumAge, maximumAge + 1);
                }
            }
            else
            {
                int AgeMinOut = 0;
                int AgeMaxOut = 0;

                GetAgeMinMax(dt, minimumAge, maximumAge, ref AgeMinOut, ref AgeMaxOut);

                if (AgeMaxOut == int.MaxValue)
                {
                    simulationCell.Age = AgeMinOut;
                }
                else
                {
                    simulationCell.Age = this.m_RandomGenerator.GetNextInteger(AgeMinOut, AgeMaxOut + 1);
                }
            }
        }

        /// <summary>
        /// Initializes the age reporting helper
        /// </summary>
        /// <remarks></remarks>
        private void InitializeAgeReportingHelper()
        {
            Debug.Assert(this.m_AgeReportingHelper == null);

            this.m_AgeReportingHelper = new AgeHelper(false, 0, 0);
            DataRow dr = this.Project.GetDataSheet(Strings.DATASHEET_AGE_TYPE_NAME).GetDataRow();

            if (dr == null)
            {
                return;
            }

            if (dr[Strings.DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME] != DBNull.Value)
            {
                if (dr[Strings.DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME] == DBNull.Value)
                {
                    this.RecordStatus(StatusType.Warning, "Age reporting freqency set without age reporting maximum.  Not reporting ages.");

                    return;
                }
            }

            if (dr[Strings.DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME] != DBNull.Value)
            {
                if (dr[Strings.DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME] == DBNull.Value)
                {
                    this.RecordStatus(StatusType.Warning, "Age reporting maximum set without age reporting frequency.  Not reporting ages.");

                    return;
                }
            }

            int f = Convert.ToInt32(dr[Strings.DATASHEET_AGE_TYPE_FREQUENCY_COLUMN_NAME], CultureInfo.InvariantCulture);
            int m = Convert.ToInt32(dr[Strings.DATASHEET_AGE_TYPE_MAXIMUM_COLUMN_NAME], CultureInfo.InvariantCulture);

            if (m < f)
            {
                this.RecordStatus(StatusType.Warning, "Age reporting maximum is less than age reporting frequency.  Not reporting ages.");

                return;
            }

            this.m_AgeReportingHelper = new AgeHelper(true, f, m);
        }

        /// <summary>
        /// Gets the age min and age max for the specified deterministic transition and age values
        /// </summary>
        /// <param name="dt">The deterministic transition</param>
        /// <param name="ageMinIn">The minimum age IN value</param>
        /// <param name="ageMaxIn">The maximum age IN value</param>
        /// <param name="ageMinOut">The minimum age OUT value</param>
        /// <param name="ageMaxOut">The maximum age OUT value</param>
        /// <remarks></remarks>
        internal static void GetAgeMinMax(DeterministicTransition dt, int ageMinIn, int ageMaxIn, ref int ageMinOut, ref int ageMaxOut)
        {
            //Normalize
            int dtagemin = Math.Min(dt.AgeMinimum, dt.AgeMaximum);
            int dtagemax = Math.Max(dt.AgeMinimum, dt.AgeMaximum);

            //This should already be normalized
            Debug.Assert(ageMinIn <= ageMaxIn);

            //If any age value is outside of the dt value, set the age value equal to the nearest dt value.
            //For example, if ageMinIn > dt.agemax the set ageMinIn = dt.agemax.

            if (ageMinIn < dtagemin)
            {
                ageMinIn = dtagemin;
            }

            if (ageMaxIn < dtagemin)
            {
                ageMaxIn = dtagemin;
            }

            if (ageMinIn > dtagemax)
            {
                ageMinIn = dtagemax;
            }

            if (ageMaxIn > dtagemax)
            {
                ageMaxIn = dtagemax;
            }

            ageMinOut = ageMinIn;
            ageMaxOut = ageMaxIn;

            Debug.Assert(ageMinOut <= ageMaxOut);
        }
    }
}
